### PR TITLE
OAuth

### PR DESCRIPTION
**Pull request by Solije**
Picking up from #6, I've added an OAuth2 client to handle communicating with Monzo's API adapted from the one at https://github.com/orcasgit/python-fitbit

MonzoOAuth2Client takes over the role of the Requests class while automatically updating the access tokens behind the scenes as they expire.

I wanted to keep backwards compatibility so the Monzo constructor still takes a naked access token.
For the automatic token refreshing, a MonzoOAuth2Client must be fully initialised and then inserted into a new Monzo object using the Monzo.from_oauth_session() class method.

I've also added some type hinting throughout the package although this can be removed if needs be.

### CODE REVIEW

**Comment by Solije** [comment]
Regarding the type hinting, I've decided it's probably best to strip it out for now.

I was having a bit of an experiment but it doesn't feel up to scratch so I'll make another commit without them in a bit.

**Comment by muyiwaolu** [comment]
This looks great. I shall take a look towards the end of next week as I've got a busy week this week.

**Comment by Solije** [comment]
Cool, no problem. I've got a couple of plans to do something similar to #7 which I'll work on in the meantime.

I'd recommend squashing some of these commits if you end up merging this as the history probably isn't that useful. I'm more used to gitlab where this can be done automatically when approving the merge. I'm not sure if github has the same option but if not then let me know and I'll manually squash them and make a new PR.

**Comment by muyiwaolu** [comment]
> I'd recommend squashing some of these commits if you end up merging this as the history probably isn't that useful. I'm more used to gitlab where this can be done automatically when approving the merge. I'm not sure if github has the same option but if not then let me know and I'll manually squash them and make a new PR.

Yes, I've set the default merge option for the repo to be `Squash and Merge`

**Comment by samdobson** [comment]
I'm going to try and have a go at implementing this in [monzo-coffee](https://github.com/samdobson/monzo-coffee) when I get the time (though I'm not sure when that will be). The interaction between this module and the database is going to become a bit tricky, I think.

I'm fairly clear on the first steps. I need to send the user to the auth URL, then once they have authenticated and returned to my app, I need to store `access_token`, `refresh_token` and `expires_at` in my database. But presumably I then need to update these database values on each token refresh?

It's making my head spin a little...

File: monzo/auth.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        """
+
+        if redirect_uri:
+            self.session.redirect_uri = redirect_uri
+
+        return(self.session.authorization_url(self.authorization_url, **kwargs))
+
+    def fetch_access_token(self, code:str, redirect_uri:str=None):
+
+        """Step 2: Given the code from fitbit from step 1, call
```
[the remark this answers is not in the document]
Comment (Solije) [reply monzo/auth.py]: Fixed in new commit. Thanks for spotting this.

File: monzo/auth.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    API_ENDPOINT = "https://api.monzo.com"
+    API_VERSION = 1
+
+    authorization_url = AUTHORIZE_ENDPOINT
+    request_token_url = "%s/oauth2/token" % API_ENDPOINT
+    access_token_url = request_token_url
+    refresh_token_url = request_token_url
+
+    def __init__(self, client_id, client_secret, access_token=None,
+            refresh_token=None, expires_at=None, refresh_cb=None,
```
[the remark this answers is not in the document]
Comment (Solije) [reply monzo/auth.py]: Hey, I'll add this to the docstring but thought I'd answer here as well as it's relevant to your other comment.

`refresh_cb` is the callback function to be called with the new token as a argument when MonzoOAuth2Client refreshes the access token. To keep your database synced, you would then just need to implement the function which updates the token in your database and pass it to MonzoOAuth2Client as `refresh_cb`, everything else should be handled automagically afterwards.

**Comment by samdobson** [comment]
As this is a fairly major overhaul of the module, we need to make sure it is straightforward to use out of the box and well-documented.

With this is mind, before this PR gets merged, there are a couple of final things I would like to see:

* Simple default token persistence (e.g. to disk/env variables)
* An end-to-end command-line usage example

Once those are in place I would say we are good to go.

**Comment by Solije** [comment]
OK, I've added a basic refresh_callback implementation to save the token and client credentials to a json file.`MonzoOAuth2Client`can also now be instansiated from such a json file directly.

It's a bit difficult to give a full run through of the authentication process as a fair amount of it depends on how the user integrates monzo-python into their application but hopefully this does the job.

**Comment by samdobson** [comment]
I'm not near a computer right now but just wanted to say a huge thanks for your work on this, it looks really thorough. I will make time tomorrow to give the code a proper test and will report back. Have a great evening!

**Comment by samdobson** [comment]
OK. Initial auth works fine:

```python
from monzo.auth import MonzoOAuth2Client
from monzo.monzo import Monzo  

oauth_client = MonzoOAuth2Client(CLIENT_ID, CLIENT_SECRET, redirect_uri=REDIRECT_URI)
auth_url, state = oauth_client.authorize_token_url()
print('Authenticate using the following URL:' + auth_url)
auth_code = input('Enter auth code:')
oauth_client.fetch_access_token(auth_code)
client = Monzo.from_oauth_session(oauth_client)
print(client.get_transactions(account_id)['transactions'])
```

But then, trying to use it:

```python
from monzo.auth import MonzoOAuth2Client
from monzo.monzo import Monzo
oauth_client = MonzoOAuth2Client(CLIENT_ID, CLIENT_SECRET, redirect_uri=REDIRECT_URI)
client = Monzo.from_oauth_session(oauth_client)
print(client.get_transactions('acc_00009Jpt6TNZmimrHGnHTV')['transactions'])
```

```python
Traceback (most recent call last):
  File "t.py", line 15, in <module>
    print(client.get_transactions(account_id)['transactions'])
  File "/home/sam/monzo-python/monzo/monzo.py", line 84, in get_transactions
    response = self.oauth_session.make_request(url, params=params)
  File "/home/sam/monzo-python/monzo/auth.py", line 127, in make_request
    **kwargs
  File "/home/sam/monzo-python/monzo/auth.py", line 105, in _request
    if d['errors'][0]['errorType'] == 'expired_token':
KeyError: 'errors'
```
I also can't see that `monzo.json` has been saved...

**Comment by samdobson** [comment]
Excellent. This now seems to be working nicely. So the TLDR OAuth quickstart, in a similar style to the existing non-oAuth one, would be something like this:

## Authenticating (this only needs to be done once)

```python3
from monzo.auth import MonzoOAuth2Client
from monzo.monzo import Monzo  

oauth_client = MonzoOAuth2Client(CLIENT_ID, CLIENT_SECRET, redirect_uri=REDIRECT_URI)
auth_url, state = oauth_client.authorize_token_url()
print('Connect your Monzo account by visiting: ' + auth_url)
auth_code = input('Enter auth code: ')
oauth_client.fetch_access_token(auth_code)
```

## Using the API

```python3
from monzo.auth import MonzoOAuth2Client
from monzo.monzo import Monzo

oauth_client = MonzoOAuth2Client.from_json()
client = Monzo.from_oauth_session(oauth_client)
print(client.get_balance(account_id))
```

**Comment by samdobson** [comment]
Right. I need to wait now until 23.00 for the magic to happen (token refresh). Assuming this all happens smoothly, I'm happy to give my seal of approval.

**Comment by Solije** [comment]
Yep, that code looks right.

If you don't want to wait you can always reach into `MonzoOAuth2Client.session.token` and mangle the `access_token` portion. That should force a refresh with the exact same logic as a real expiry.

**Comment by samdobson** [comment]
I tried mangling the access token in `monzo.json`, but no joy: ```monzo.errors.BadRequestError: Token is invalid```

Have to run now but will check back later tonight.

**Comment by Solije** [comment]
> ```python
> Traceback (most recent call last):
>   File "t.py", line 15, in <module>
>     print(client.get_transactions(account_id)['transactions'])
>   File "/home/sam/monzo-python/monzo/monzo.py", line 84, in get_transactions
>     response = self.oauth_session.make_request(url, params=params)
>   File "/home/sam/monzo-python/monzo/auth.py", line 127, in make_request
>     **kwargs
>   File "/home/sam/monzo-python/monzo/auth.py", line 105, in _request
>     if d['errors'][0]['errorType'] == 'expired_token':
> KeyError: 'errors'
> ```

I should really read error messages more carefully in future. Not my proudest moment, sorry about that. This should be handled properly now

**Comment by Solije** [comment]
> I tried mangling the access token in `monzo.json`, but no joy: `monzo.errors.BadRequestError: Token is invalid`
> 
> Reducing `expires_at` results in `oauthlib.oauth2.rfc6749.errors.InvalidClientError: (invalid_client) Provide client credentials (client ID and client secret)`
> 
> Have to run now but will check back later tonight.

I had a bit of a go mangling access tokens just now. I found that `MonzoOAuth2Client` can only recover from corruptions in certain positions in the string and you otherwise get that `monzo.errors.BadRequestError: Token is invalid` error.

Some areas of the key remain constant between refreshes which I think is key to the validity issue

**Comment by samdobson** [comment]
My bad, it actually expired at 23.00 *today* (expiry length was much longer than expected: 30hrs). Unfortunately the token refresh failed:

```
sam@hetserve:~/monzo-python$ python3 t.py 
Traceback (most recent call last):
  File "/home/sam/.local/lib/python3.6/site-packages/requests_oauthlib/oauth2_session.py", line 330, in request
    http_method=method, body=data, headers=headers)
  File "/home/sam/.local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 196, in add_token
    raise TokenExpiredError()
oauthlib.oauth2.rfc6749.errors.TokenExpiredError: (token_expired) 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "t.py", line 26, in <module>
    trans = client.get_transactions('acc_00009Jpt6TNZmimrHGnHTV')['transactions']
  File "/home/sam/monzo-python/monzo/monzo.py", line 84, in get_transactions
    response = self.oauth_session.make_request(url, params=params)
  File "/home/sam/monzo-python/monzo/auth.py", line 127, in make_request
    **kwargs
  File "/home/sam/monzo-python/monzo/auth.py", line 99, in _request
    response = self.session.request(method, url, **kwargs)
  File "/home/sam/.local/lib/python3.6/site-packages/requests_oauthlib/oauth2_session.py", line 343, in request
    self.auto_refresh_url, auth=auth, **kwargs
  File "/home/sam/.local/lib/python3.6/site-packages/requests_oauthlib/oauth2_session.py", line 309, in refresh_token
    self.token = self._client.parse_request_body_response(r.text, scope=self.scope)
  File "/home/sam/.local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 411, in parse_request_body_response
    self.token = parse_token_response(body, scope=scope)
  File "/home/sam/.local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 379, in parse_token_response
    validate_token_parameters(params)
  File "/home/sam/.local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 386, in validate_token_parameters
    raise_from_error(params.get('error'), params)
  File "/home/sam/.local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/errors.py", line 415, in raise_from_error
    raise cls(**kwargs)
oauthlib.oauth2.rfc6749.errors.InvalidClientError: (invalid_client) Provide client credentials (client ID and client secret)
```

**Comment by muyiwaolu** [comment]
This is looking pretty good. I'm going to give it a proper sit down review and go through a plan to merge this in. This is, semantically, a minor update, right @Solije @samdobson?

Also, great work! Thanks for the PRs.

**Comment by samdobson** [comment]
@muyiwaolu I believe it is. Other than adding the ability to use OAuth, the original module usage and functionality I think remains unchanged. @Solije - correct me if I'm wrong.

**Comment by Solije** [comment]
The only real change other than the OAuth capabilities is that I added the `update_transaction_metadata` endpoint to match with #12. Other than that, yeah, no new functionality apart from OAuth.

**Comment by Solije** [comment]
I think that I know what's the issue with the client id when refreshing the token but it throws up another question.

I was a bit silly and didn't check if Fitbit and Monzo were using a different implementation of the refresh step of the OAuth procedure. Having a bit more of a look at the two sets of docs, Fitbit expected the client id and secret to be given to it using HTTP Basic auth while Monzo just wants them added to the request as parameters.

The only issue is that I would imagine that if this was the case, then forcing a refresh should cause a similar error to the one we get after the token expires. I've had no issues with forced refreshes so I'm not entirely sure how that affects things.

I'm going to be busy for the next few days but I'd like to have a little play with how I attach the id and secret to that refresh call.

**Comment by Solije** [comment]
I've just made the changes as above and have got a token which is expiring in the morning. I'll let you guys know if it crashes and burns or not

**Comment by Solije** [comment]
Found the issue. Will have the fix up soon

**Comment by Solije** [comment]
For some reason I had that `auto_refresh_url` parameter set up when initialising the oauth session, which meant that the `refresh_token` method was being bypassed automatically. However, as there wasn't a corresponding parameter set for the client id and secret to be passed when refreshing, it was failing.

I've changed it to manually catch the exception for an expired token and direct it to `self.refresh_token()` now. It worked on the expired token I got this morning but obviously I can't test much more for now.

**Comment by Solije** [comment]
Second token expiry has passed without a hitch. @muyiwaolu, I think this is ready for primetime.

**Comment by samdobson** [comment]
Awesome work! :)

**Comment by muyiwaolu** [comment]
@Solije woop! Rebase this baby, and we're good to go!

**Pull request merged**